pkg/scheduler/schedule_one.go

```diff
@@ -879,6 +895,11 @@ func (sched *Scheduler) handleSchedulingFailure(ctx context.Context, fwk framewo
 		klog.ErrorS(err, "Error scheduling pod; retrying", "pod", klog.KObj(pod))
 	}
 
+	if len(unschedulablePlugins) != 0 {
+		// Inject UnschedulablePlugins to PodInfo, which will be used later for moving Pods between queues efficiently.
+		podInfo.UnschedulablePlugins = unschedulablePlugins
```

Comment:
This is the change that was meant to ensure that `Reserve` failures for DRA add the plugin to `UnschedulablePlugins`, right?

I had to add some additional debug output, which then gave me the smoking gun that explains why some pods randomly got stuck:
```
    logger.go:85: E0512 09:45:41.637696   35146 schedule_one.go:895] "Error scheduling pod; retrying" err="running Reserve plugin \"DynamicResources\": waiting for resource driver to allocate resource" pod="init/test-dra9gjsz"
    logger.go:85: I0512 09:45:41.637718   35146 schedule_one.go:900] "Unschedulable plugins after failure" pod="init/test-dra9gjsz" unschedulablePlugins={"":{}}
    logger.go:85: I0512 09:45:41.637751   35146 scheduling_queue.go:649] "Pod moved to an internal scheduling queue" pod="init/test-dra9gjsz" event="ScheduleAttemptFailure" queue="Unschedulable"
...
    logger.go:85: I0512 09:45:41.641337   35146 scheduling_queue.go:909] "Event is not of a kind that some unschedulable plugin waits for" pod="init/test-dra9gjsz" unschedulablePlugins={"":{}} event="ResourceClaimUpdate"

```

`Reserve` has failed, but the DRA plugin is not getting added. Instead, there is an empty entry in the set (`""`).

Later, `movePodsToActiveOrBackoffQueue` ignores the event that would have enabled scheduling.

This seems to come from `sets.New(sts.FailedPlugin())` when running reserve plugins in `schedulingCycle`, which in turn comes from `RunReserve`.

With the following changes, I get pods scheduled more reliably:
```patch
diff --git a/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go b/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
index 32ee6aa39f8..7c2cfbe43e7 100644
--- a/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
+++ b/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
@@ -263,6 +263,12 @@ func (pl *dynamicResources) Requeue(ctx context.Context, p *v1.Pod, event *frame
 		return framework.PodNotAffected
 	}
 
+	if event.ActionType&framework.Delete != 0 {
+		// isSchedulableAfterClaimChange and isSchedulableAfterPodSchedulingChange expect
+		// the deleted object in oldObj.
+		oldObj, obj = obj, oldObj
+	}
+
 	switch event.Resource {
 	case framework.ResourceClaim:
 		return pl.isSchedulableAfterClaimChange(oldObj, obj, p)
diff --git a/pkg/scheduler/framework/runtime/framework.go b/pkg/scheduler/framework/runtime/framework.go
index fb89ee7d0f4..83a94f0cbfd 100644
--- a/pkg/scheduler/framework/runtime/framework.go
+++ b/pkg/scheduler/framework/runtime/framework.go
@@ -1170,7 +1170,9 @@ func (f *frameworkImpl) RunReservePluginsReserve(ctx context.Context, state *fra
 		if !status.IsSuccess() {
 			err := status.AsError()
 			klog.ErrorS(err, "Failed running Reserve plugin", "plugin", pl.Name(), "pod", klog.KObj(pod))
-			return framework.AsStatus(fmt.Errorf("running Reserve plugin %q: %w", pl.Name(), err))
+			status := framework.AsStatus(fmt.Errorf("running Reserve plugin %q: %w", pl.Name(), err))
+			status.SetFailedPlugin(pl.Name())
+			return status
 		}
 	}
 	return nil
diff --git a/pkg/scheduler/internal/queue/scheduling_queue.go b/pkg/scheduler/internal/queue/scheduling_queue.go
index bfe7ded53c3..0e4431691c6 100644
--- a/pkg/scheduler/internal/queue/scheduling_queue.go
+++ b/pkg/scheduler/internal/queue/scheduling_queue.go
@@ -906,12 +906,14 @@ func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(podInfoList []*framework.
 		// either there is some abnormal error, or scheduling the pod failed by plugins other than PreFilter, Filter and Permit.
 		// In that case, it's desired to move it anyways.
 		if len(pInfo.UnschedulablePlugins) != 0 && !p.podMatchesEvent(pInfo, event) {
+			klog.V(6).InfoS("Event is not of a kind that some unschedulable plugin waits for", "pod", klog.KObj(pInfo.Pod), "unschedulablePlugins", pInfo.UnschedulablePlugins, "event", event.Label)
 			continue
 		}
 
 		schedulingHint := p.isPodWorthRequeuing(ctx, pInfo, &event, oldObj, newObj)
 		if schedulingHint == framework.PodNotAffected {
 			// This Pod isn't worth putting to activeQ or backoffQ.
+			klog.V(6).InfoS("Event is not making pod schedulable", "pod", klog.KObj(pInfo.Pod), "event", event.Label)
 			continue
 		}
 
diff --git a/pkg/scheduler/schedule_one.go b/pkg/scheduler/schedule_one.go
index e7c49781372..b1c8611871d 100644
--- a/pkg/scheduler/schedule_one.go
+++ b/pkg/scheduler/schedule_one.go
@@ -897,6 +897,7 @@ func (sched *Scheduler) handleSchedulingFailure(ctx context.Context, fwk framewo
 
 	if len(unschedulablePlugins) != 0 {
 		// Inject UnschedulablePlugins to PodInfo, which will be used later for moving Pods between queues efficiently.
+		klog.V(6).InfoS("Unschedulable plugins after failure", "pod", klog.KObj(podInfo.Pod), "unschedulablePlugins", unschedulablePlugins)
 		podInfo.UnschedulablePlugins = unschedulablePlugins
 	}
```

Reply:
@pohly Great catch. Thanks for the investigation 👍